lib/fizzy/execute.cpp

```diff
@@ -213,8 +224,57 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
             goto end;
         case Instr::nop:
             break;
+        case Instr::block:
+        {
+            const auto arity = read<uint8_t>(immediates);
+
+            if (arity != 0)
+            {
+                throw std::runtime_error(
+                    "block instruction with non-empty result type not implemented yet");
+            }
+
+            const auto target_pc = read<uint32_t>(immediates);
+            const auto target_imm = read<uint32_t>(immediates);
+            LabelContext label{code.instructions.data() + target_pc,
+                code.immediates.data() + target_imm, arity, stack.size()};
+            labels.emplace_back(label);
+            break;
+        }
+        case Instr::loop:
+        {
+            LabelContext label{pc - 1, immediates, 0, stack.size()};  // Target this instruction.
+            labels.push_back(label);
+            break;
+        }
         case Instr::end:
-            goto end;
+        {
+            if (!labels.empty())
+                labels.pop_back();
+            else
+                goto end;
+            break;
+        }
+        case Instr::br:
+        case Instr::br_if:
+        {
+            const auto label_idx = read<uint32_t>(immediates);
+
+            // Check condition for br_if.
+            if (instruction == Instr::br_if && static_cast<uint32_t>(stack.pop()) == 0)
```

Comment:
Why this `static_cast`? And what if `br_if` checks result of `i64.add`, will it ignore high bytes?

Reply:
![Screenshot 2020-01-16 at 15 21 09](https://user-images.githubusercontent.com/20340/72537316-d38f4d80-3873-11ea-91e8-f9b718003c8f.png)

Reply:
Maybe a comment in the code would be useful.

Reply:
This is complicated and affects many other instructions implementations.

There are two options when putting `i32` values in our implementation of the stack with `uint64_t` types:

1. Zero-extend to `uint64_t`. This is 64-bit memory store instruction.
2. Only store 32-bits, leaving garbage in the high word on the stack.

We are currently doing 1, mostly because this is much easier to do in C/C++ than 2.

So here the `static_cast` is not needed because the high bits are going to be zeros anyway. The `static_cast` would be required if we implement option 2.

Nevertheless, using `static_cast` emits an assembly instruction for 32-bit load instead of 64-bit load. Probably not any faster one (no way 64-bit one can cross cache line here). But you saved one byte in the final code size.